pennylane/interfaces/execution.py

```diff
@@ -740,10 +763,20 @@ def device_gradient_fn(inner_tapes, **gradient_kwargs):
         # within execute_and_gradients, so providing a gradient_fn
         # in this case would have ambiguous behaviour.
         raise ValueError("Gradient transforms cannot be used with grad_on_execution=True")
+    elif interface in jpc_interfaces:
+        jpc = TransformJacobianProducts(execute_fn, gradient_fn, gradient_kwargs)
+        for _ in range(1, max_diff):
+            ml_boundary_execute = _get_ml_boundary_execute(interface, _grad_on_execution)
+            execute_fn = partial(ml_boundary_execute, execute_fn=execute_fn, jpc=jpc)
+            jpc = TransformJacobianProducts(execute_fn, gradient_fn, gradient_kwargs)
 
     ml_boundary_execute = _get_ml_boundary_execute(interface, _grad_on_execution)
-    results = ml_boundary_execute(
-        tapes, device, execute_fn, gradient_fn, gradient_kwargs, _n=1, max_diff=max_diff
-    )
+
+    if interface in jpc_interfaces:
+        results = ml_boundary_execute(tapes, execute_fn, jpc)
+    else:
+        results = ml_boundary_execute(
+            tapes, device, execute_fn, gradient_fn, gradient_kwargs, _n=1, max_diff=max_diff
+        )

```

Comment:
As far as I can tell, the integration of the new interface branch is correct :) 
This is a rather involved function, though, so I'd be happy for more eyes to check this and maybe for more dev comments about the strategy.
One detail that caught my eye was that `grad_on_execution`, `_grad_on_execution` and `config.grad_on_execution` are a bit hard to track throughout this nested if structure. In particular, I'm not sure we need `_grad_on_execution`? Why can't we overwrite `grad_on_execution` depending on the branch of the if clauses?

Reply:
So `config.grad_on_execution` is just for the new device.  I'm actually not entirely sure about the difference between `grad_on_execution` and `_grad_on_execution`.

Reply:
I see :+1: Yeah this seems like a grown structure :sweat_smile: 
I'm not sure it's worth touching it...